File: app/src/main/java/com/pr0gramm/app/parcel/TagListParceler.java
package com.pr0gramm.app.parcel;

import android.os.Parcel;
import android.os.Parcelable;

import com.pr0gramm.app.api.pr0gramm.Api;
import com.pr0gramm.app.parcel.core.Parceler;

import java.util.List;

/**
 */
public class TagListParceler extends Parceler<List<Api.Tag>> {
    public TagListParceler(List<Api.Tag> values) {
        super(values);
    }

    @SuppressWarnings("unused")
    private TagListParceler(Parcel parcel) {
        super(parcel);
    }

    public static final Parcelable.Creator<TagListParceler> CREATOR =
            new LambdaCreator<>(TagListParceler::new, TagListParceler[]::new);
}

File: app/src/main/java/com/pr0gramm/app/parcel/TagListParceler.kt
package com.pr0gramm.app.parcel

import com.pr0gramm.app.api.pr0gramm.Api

/**
 */
class TagListParceler(val tags: List<Api.Tag>) : Freezable {
    override fun freeze(sink: Freezable.Sink) = with(sink) {
        sink.writeValues(tags.size) { idx ->
            val tag = tags[idx]
            writeLong(tag.id)
            writeFloat(tag.confidence)
            writeString(tag.tag)
        }
    }

    companion object : Unfreezable<TagListParceler> {
        @JvmField
        val CREATOR = parcelableCreator()

        override fun unfreeze(source: Freezable.Source): TagListParceler = with(source) {
            val tags = readValues {
                val id = readLong()
                val confidence = readFloat()
                val tag = readString()
                Api.Tag(id, confidence, tag)
            }

            return TagListParceler(tags)
        }
    }
}
